packages/tasit-action/src/TasitAction.js

```diff
@@ -19,117 +22,370 @@ class Utils {
     }
 
-    if (!callback || typeof callback !== "function") {
-      throw new Error(`Cannot subscribe without a function`);
+    if (!listener || typeof listener !== "function") {
+      throw new Error(`Cannot listening without a function`);
     }
 
-    const blockHeight = 1;
-    const tx = await this.#txPromise;
+    if (eventName === "error") {
+      this.addErrorListener(listener);
+      return;
+    }
 
-    this.#provider.on(tx.hash, async receipt => {
-      const { confirmations } = receipt;
-      const message = {
-        data: {
-          confirmations: confirmations,
-        },
-      };
+    this.#tx = await this.#txPromise;
 
+    let wrappedListener = async blockNumber => {
       try {
-        await callback(message);
+        const receipt = await this.#provider.getTransactionReceipt(
+          this.#tx.hash
+        );
+        const txConfirmed = receipt !== null;
+
+        if (txConfirmed) {
+          this.setEventInfo("block", wrappedListener, {
+            hasBeenConfirmed: true,
+          });
+        } else {
+          const { hasBeenConfirmed } = this.getEventInfo(
+            "block",
+            wrappedListener
+          );
+          if (hasBeenConfirmed)
+            this.emitErrorEvent(
+              new Error(`Your message has been included in an uncle block.`)
+            );
+
+          return;
+        }
+
+        const { confirmations } = receipt;
+        const message = {
+          data: {
+            confirmations: confirmations,
+          },
+        };
+
+        await listener(message);
       } catch (error) {
-        throw new Error(`Callback function with error: ${error.message}`);
+        this.emitErrorEvent(
+          new Error(`Callback function with error: ${error.message}`)
+        );
       }
-    });
+    };
+
+    this.addListener(eventName, "block", listener, wrappedListener);
   };
 
-  removeAllListeners = async () => {
+  // Tech debt
+  // This method avoid duplicated nonce generation when rapid succession of several transactions
+  // See: https://github.com/ethereumbook/ethereumbook/blob/04f66ae45cd9405cce04a088556144be11979699/06transactions.asciidoc#keeping-track-of-nonces
+  // How we'll should keeping track of nonces?
+  waitForMessage = async () => {
     const tx = await this.#txPromise;
-    this.#provider.removeAllListeners(tx.hash);
+
+    await this.#provider.waitForTransaction(tx.hash);
   };
 }
 
-class EventSubscription {
-  #eventNames;
+class ContractSubscription extends Subscription {
+  #contract;
   #provider;
+  #eventNames;
 
-  constructor(eventNames, provider) {
-    this.#eventNames = eventNames;
+  // Why eventNames shoud be passed here?
+  constructor(contract, provider, eventNames) {
+    eventNames.forEach(eventName => {
+      if (contract.interface.events[eventName] === undefined)
+        throw new Error(`Event '${eventName}' not found.`);
+    });
+
+    super(contract);
```

Comment:
the contract is connected w/ a provider already.

Reply:
Ah ok. I ask because in TransactionSubscription it does `super(provider);`

Does TransactionSubscription not need a contract? I wonder if more parallel structure could make sense in this regard. Maybe not though.

Reply:
Both `provider` and `contract` implements EventEmitter API. I'm calling both as `emitter` on SuperClass. As `TransactionSubscription` encapsulates normal transaction events, it's not necessary to be aware of the contract object. I'm not sure, but I think that way we achieve +cohesion and -coupling. What do you think?

Reply:
I would argue that since transactions often involve an address that's a contract, it could still sometimes make sense to be aware of the contract object. But I'd have to think about this more.

This makes sense - I would say this is the type of thing that's odd enough on first glance that it's worth adding a line or two of comments up near the constructor of each explaining the different choice for each

All in all I'm on board with the current approach!